doc/sphinx/source/driver_design.rst

```diff
@@ -0,0 +1,179 @@
+=========================================
+How to write a new device driver for Pocl
+=========================================
+
+The first required steps are adding the CMake infrastructure, and
+making the driver known to device initialization. Required CMake
+changes are: a boolean option to enable the driver (optionally
+autodetected at CMake time), adding the bool to config.h.in.cmake,
+finding all required libraries, linking & compilation options,
+and correctly linking the driver depending on ``ENABLE_LOADABLE_DRIVERS``
+(these can be copied from existing drivers). To make the driver known
+to device initialization in devices.c, add necessary #includes, and
+the new driver name to ``pocl_device_types`` and ``pocl_devices_init_ops``.
+
+Setting device properties
+-----------------------------------
+
+In the driver's init callback, the driver should discover and set up each device's properties.
+These are stored as members of ``struct _cl_device_id``, e.g. ``max_work_item_dimensions`` or
+``max_compute_units``. Some of them are internal to PoCL, e.g. ``device_side_printf`` is a bool
+that changes the processing of the 'printf' function in PoCL's LLVM passes. There are some helpers
+for setting up properties; for setting up versioned (OpenCL 3.0) properties, see ``lib/CL/devices/common.c``;
+the same file also has ``pocl_init_default_device_infos`` which sets up some reasonable defaults.
+By default the entire struct is memset to zero, so all properties are zero / false.
+
+
+PoCL's internal driver API
+-----------------------------------
+
+Next step is to implement PoCL's internal driver API operations (callbacks) for the device.
```

Comment:
maybe add (user space) before 'driver' to make it clear we do not have kernel space drivers in PoCL